File: backend/app/schemas/client_email.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel


EmailDirection = Literal["sent", "received", "unknown"]


class ClientEmailAttachmentRead(BaseModel):
    document_id: int
    original_filename: str | None
    content_type: str
    file_size: int


class ClientEmailRead(BaseModel):
    id: int
    external_id: str
    message_id: str
    thread_id: str | None
    direction: EmailDirection
    message_at: datetime | None
    from_name: str | None
    from_address: str | None
    to_addresses: list[str]
    cc_addresses: list[str]
    subject: str | None
    body_text: str | None
    source_url: str | None
    attachment_count: int
    attachments: list[ClientEmailAttachmentRead]
    created_at: datetime


class ClientEmailPage(BaseModel):
    items: list[ClientEmailRead]
    total: int
    skip: int
    limit: int

File: frontend/lib/features/clients/domain/client_email.dart
class ClientEmailAttachment {
  const ClientEmailAttachment({
    required this.documentId,
    required this.contentType,
    required this.fileSize,
    this.originalFilename,
  });

  final int documentId;
  final String? originalFilename;
  final String contentType;
  final int fileSize;

  String get displayName => originalFilename ?? 'Załącznik #$documentId';
}

enum ClientEmailDirection { sent, received, unknown }

class ClientEmail {
  const ClientEmail({
    required this.id,
    required this.externalId,
    required this.messageId,
    required this.direction,
    required this.toAddresses,
    required this.ccAddresses,
    required this.attachmentCount,
    required this.attachments,
    required this.createdAt,
    this.ignored = false,
    this.threadId,
    this.messageAt,
    this.fromName,
    this.fromAddress,
    this.subject,
    this.bodyText,
    this.sourceUrl,
  });

  final int id;
  final String externalId;
  final String messageId;
  final String? threadId;
  final ClientEmailDirection direction;
  final DateTime? messageAt;
  final String? fromName;
  final String? fromAddress;
  final List<String> toAddresses;
  final List<String> ccAddresses;
  final String? subject;
  final String? bodyText;
  final String? sourceUrl;
  final int attachmentCount;
  final List<ClientEmailAttachment> attachments;
  final DateTime createdAt;
  final bool ignored;

  String get displaySubject => subject ?? '(bez tematu)';
}
